pkg/servicemesh/controller/controller.go

```diff
@@ -70,78 +75,43 @@ func (smmrc *serviceMeshMemberRollController) Start(stop chan struct{}) {
 }
 
 func (smmrc *serviceMeshMemberRollController) Register(listener MemberRollListener) {
-	smmrc.informer.AddEventHandler(smmrc.newServiceMeshMemberRollListener(listener))
-}
-
-func (smmrc *serviceMeshMemberRollController) getNamespaces(namespaces []string) []string {
-	result := append([]string(nil), namespaces...)
-
-	found := false
-	for _, namespace := range namespaces {
-		if namespace == smmrc.namespace {
-			found = true
-			break
-		}
-	}
-	if !found {
-		result = append(result, smmrc.namespace)
-	}
-	return result
-}
-
-func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(listener MemberRollListener) cache.ResourceEventHandler {
-	return &serviceMeshMemberRollListener{
-		smmrc:             smmrc,
-		listener:          listener,
-		currentNamespaces: smmrc.getNamespaces(nil),
-	}
-}
-
-type serviceMeshMemberRollListener struct {
-	smmrc             *serviceMeshMemberRollController
-	listener          MemberRollListener
-	currentNamespaces []string
-}
-
-func (smmrl *serviceMeshMemberRollListener) checkEquality(lhs, rhs []string) bool {
-	if (lhs == nil) != (rhs == nil) {
-		return false
-	}
-	if len(lhs) != len(rhs) {
-		return false
-	}
-	for n, val := range lhs {
-		if val != rhs[n] {
-			return false
-		}
-	}
-	return true
+	smmrc.lock.Lock()
+	defer smmrc.lock.Unlock()
+	smmrc.listeners = append(smmrc.listeners, listener)
+	// push initial state
+	go listener.UpdateNamespaces(smmrc.currentNamespaces.UnsortedList())
 }
 
-func (smmrl *serviceMeshMemberRollListener) updateNamespaces(operation string, memberRollName string, members []string) {
-	if smmrl.smmrc.memberRollName != memberRollName {
+func (smmrc *serviceMeshMemberRollController) updateNamespaces(operation string, memberRollName string, members []string) {
+	smmrc.lock.Lock()
+	defer smmrc.lock.Unlock()
+	if smmrc.memberRollName != memberRollName {
 		log.Infof("ServiceMeshMemberRoll using incorrect name %v, ignoring", memberRollName)
 	} else {
-		namespaces := smmrl.smmrc.getNamespaces(members)
-		if !smmrl.checkEquality(smmrl.currentNamespaces, namespaces) {
-			smmrl.currentNamespaces = namespaces
-			log.Infof("ServiceMeshMemberRoll %v %s, namespaces now %q", memberRollName, operation, smmrl.currentNamespaces)
-			smmrl.listener.UpdateNamespaces(smmrl.currentNamespaces)
+		namespaces := sets.NewString(members...)
+		namespaces.Insert(smmrc.namespace)
+		if !smmrc.currentNamespaces.Equal(namespaces) {
+			smmrc.currentNamespaces = namespaces
+			log.Infof("ServiceMeshMemberRoll %v %s, namespaces now %q", memberRollName, operation, smmrc.currentNamespaces.List())
+			for _, listener := range smmrc.listeners {
+				// give everyone their own copy
+				go listener.UpdateNamespaces(smmrc.currentNamespaces.UnsortedList())
+			}
```

Comment:
Would it be preferable to use a sorted list?  This might provide some consistency when testing.

Reply:
I'm not sure that's a good idea, as it would sort the list every time this is called. If we really want that, we should probably cache the sorted list every time the set is updated